libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1564,18 +1564,75 @@ bool ExpressionCompiler::visit(IndexAccess const& _indexAccess)
 			break;
 		case DataLocation::CallData:
-			//@todo if we implement this, the value in calldata has to be added to the base offset
-			solUnimplementedAssert(!arrayType.baseType()->isDynamicallySized(), "Nested arrays not yet implemented.");
-			if (arrayType.baseType()->isValueType())
-				CompilerUtils(m_context).loadFromMemoryDynamic(
-					*arrayType.baseType(),
-					true,
-					!arrayType.isByteArray(),
-					false
-				);
+			if (arrayType.baseType()->isDynamicallyEncoded())
+			{
+				// stack layout: <base_ref> <length> <index>
+				ArrayUtils(m_context).accessIndex(arrayType, true, true);
+				// stack layout: <base_ref> <ptr_to_tail>
+
+				unsigned int baseEncodedSize = arrayType.baseType()->calldataEncodedSize();
+				solAssert(baseEncodedSize > 1, "");
+
+				// returns the absolute offset of the accessed element in "base_ref"
+				m_context.appendInlineAssembly(Whiskers(R"({
+					let rel_offset_of_tail := calldataload(ptr_to_tail)
+					if iszero(slt(rel_offset_of_tail, sub(sub(calldatasize(), base_ref), sub(<neededLength>, 1)))) { revert(0, 0) }
+					base_ref := add(base_ref, rel_offset_of_tail)
+				})")("neededLength", toCompactHexWithPrefix(baseEncodedSize)).render(), {"base_ref", "ptr_to_tail"});
+				// stack layout: <absolute_offset_of_tail> <garbage>
+
+				if (!arrayType.baseType()->isDynamicallySized())
+				{
+					m_context << Instruction::POP;
+					// stack layout: <absolute_offset_of_element>
+					solAssert(
+						arrayType.baseType()->category() == Type::Category::Struct ||
+						arrayType.baseType()->category() == Type::Category::Array,
+						"Invalid dynamically encoded base type on array access."
+					);
+				}
+				else
+				{
+					auto const* baseArrayType = dynamic_cast<ArrayType const*>(arrayType.baseType().get());
+					solAssert(!!baseArrayType, "Invalid dynamically sized type.");
+					unsigned int calldataStride = baseArrayType->calldataStride();
+					solAssert(calldataStride > 0, "");
+
+					// returns the absolute offset of the accessed element in "base_ref"
+					// and the length of the accessed element in "ptr_to_length"
+					m_context.appendInlineAssembly(
+						Whiskers(R"({
+							length := calldataload(base_ref)
+							base_ref := add(base_ref, 0x20)
+							if gt(length, 0xffffffffffffffff) { revert(0, 0) }
+							if sgt(base_ref, sub(calldatasize(), mul(length, <calldataStride>))) { revert(0, 0) }
+						})")("calldataStride", toCompactHexWithPrefix(calldataStride)).render(),
+						{"base_ref", "length"}
+					);
+					// stack layout: <absolute_offset_of_element> <length_of_element>
+				}
+			}
+			else
+			{
+				ArrayUtils(m_context).accessIndex(arrayType, true);
+				if (arrayType.baseType()->isValueType())
+					CompilerUtils(m_context).loadFromMemoryDynamic(
+						*arrayType.baseType(),
+						true,
+						!arrayType.isByteArray(),
+						false
+					);
+				else
+					solAssert(
```

Comment:
I think ``accessIndex`` above *should* have performed all necessary bounds checks in this case already.

Reply:
Yep: https://github.com/ethereum/solidity/pull/5996/files#r257742080